### PR TITLE
parameters validation

### LINKED ISSUE #184: Client side parameter validation
Validating the parameters provided in the command before it goes to the API

### PR DESCRIPTION
**Pull request by ashi31**
This is to verify the validity of input parameters for the commands before it reaches the API. The basic validation will be done in the beginning to prevent unnecessary API calls.

ping :
valididty of peerId, i.e. length & prefix

addbootstrap :
len of peers provided
if peers start with "/"

removebootstrap :
len of peers provided
if peers start with "/"

createdid :
didType (between 0 and 4)
if required files are present

addquorum :
if the file is available
if the file is empty

setupquorum :
valididty of did by checking length and prefix
if did is empty, asking the user for did

generatetestrbt :
if did is empty, asking the user for did
valididty of did by checking length and prefix
checking for numTokens value >0

transferrbt :
checking sender address, if empty asking user to enter
checking receiver address, if empty asking user to enter
checking if the adress has just did
checking for the validity of both sender and receiver DID
if transtype is 1 or 2
if rbt amount is not less than 5 decimal places

getaccountinfo :
if did is empty, asking the user for did
valididty of did by checking length and prefix

dumptokenchain :
if token is empty, asking the user for token
valididty of token by checking length and prefix

registerdid / setupdid :
if did is empty, asking the user for did
valididty of did by checking length and prefix

migratenode :
didType (between 0 and 4)

createdatatoken :
if did is empty, asking the user for did
valididty of did by checking length and prefix

commitdatatoken :
if did is empty, asking the user for did
valididty of did by checking length and prefix

gettxndetails :
checking if any of the values (did, txnId, transComment) are provided
if did is provided, valididty check of did by checking length and prefix

publishcontract :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix
checking validity of DID
checking if publishtype is 0 or 1

subscribecontract :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix

deploysmartcontract :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix
checking deployerdid by checking length and prefix
validating rbt amount to be committed
checking if trans type is 1 or 2

generatesmartcontract :
if did is empty, asking the user for did
valididty of did by checking length and prefix
checking if binary code file, raw code file and schema file are provided

fetchsmartcontract :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix

dumpsmartcontracttokenchain :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix

executesmartcontract :
if sc token is empty, asking the user for token
valididty of sc token by checking length and prefix
checking executerdid by checking length and prefix
checking if trans type is 1 or 2
checking if data to be executed is provided

checkquorumstatus :
if quorum address is empty, asking the user for quorum address
checking if address has 2 parts
check peerID and did validity

addexplorer :
length of links provided

remove explorer :
length of links provided

addpeerdetails :
if peerID is empty, asking the user for peerID
valididty of peerID by checking length and prefix
if did is empty, asking the user for did
valididty of did by checking length and prefix
didType (between 0 and 4)

### CODE REVIEW

File: command/data_token.go [hunk 1]
```diff
@@ -5,13 +5,27 @@ import (
...
+	if cmd.did == "" {
+		cmd.log.Info("DID cannot be empty")
+		fmt.Print("Enter DID : ")
+		_, err := fmt.Scan(&cmd.did)
+		if err != nil {
+			cmd.log.Error("Failed to get DID")
+			return
+		}
+	}
+	if !strings.HasPrefix(cmd.did, "bafybmi") || len(cmd.did) < 59 {
```
Comment (harirubix) [inline command/data_token.go:25]: The did length check should be equal to 59 ?

File: command/diagnostic.go [hunk 2]
```diff
@@ -133,6 +134,20 @@ func tcMarshal(str string, m interface{}) (string, error) {
...
+	if cmd.token == "" {
+		cmd.log.Info("token id cannot be empty")
+		fmt.Print("Enter Token Id : ")
+		_, err := fmt.Scan(&cmd.token)
+		if err != nil {
+			cmd.log.Error("Failed to get Token ID")
+			return
+		}
+	}
+	if len(cmd.token) < 46 || !strings.HasPrefix(cmd.token, "Qm") {
```
Comment (harirubix) [inline command/diagnostic.go:146]: The token check should be equal to 46 as well.

File: command/ping.go [hunk 3]
```diff
@@ -14,6 +23,28 @@ func (cmd *Command) ping() {
 }
 
 func (cmd *Command) checkQuorumStatus() {
+	if cmd.quorumAddr == "" {
```
Comment (harirubix) [inline command/ping.go:26]: We should change from `period.did` to `did` if we haven't already for quorum status check

File: command/transfer.go [hunk 4]
```diff
@@ -1,10 +1,47 @@
...
+	if cmd.receiverAddr == "" {
+		cmd.log.Info("Receiver address cannot be empty")
+		fmt.Print("Enter Receiver DID : ")
+		_, err := fmt.Scan(&cmd.receiverAddr)
+		if err != nil {
+			cmd.log.Error("Failed to get Receiver DID")
+			return
+		}
+	}
+	if strings.Contains(cmd.senderAddr, ".") || strings.Contains(cmd.receiverAddr, ".") {
```
Comment (harirubix) [inline command/transfer.go:29]: Not necessarily just `.` . The better check will be to make sure string is alphanumeric.

File: command/transfer.go [hunk 5]
```diff
@@ -1,10 +1,47 @@
...
+		cmd.log.Error("Invalid sender or receiver address. Please provide valid DID")
+		return
+	}
+	if !strings.HasPrefix(cmd.senderAddr, "bafybmi") || len(cmd.senderAddr) < 59 || !strings.HasPrefix(cmd.receiverAddr, "bafybmi") || len(cmd.receiverAddr) < 59 {
+		cmd.log.Error("Invalid sender or receiver DID")
+		return
+	}
+	if cmd.rbtAmount == 0.0 || cmd.rbtAmount < 0.00001 {
```
Comment (harirubix) [inline command/transfer.go:37]: `cmd.rbtAmount < 0.00001` will stop any value from 0.0 & less . So the check  `cmd.rbtAmount == 0.0` is redundant

**Review by thewebchap — CHANGES_REQUESTED** [review]

**Review by thewebchap — APPROVED** [review]

**Pull request merged**